### ISSUE TITLE
Is this a typeof bug?

### ISSUE THREAD

**Issue by jzakiya** [crystal-lang/crystal#9069]
Using 0.34 on Linux.

I have this code which works correctly setting n always to a ``BigInt``.

```
def primep3?(n)                           # P3 Prime Generator primality test
  # P3 = 6*k + {5, 7}                     # P3 primes candidates (pc) sequence
  n = n.to_big_i 
  #n = typeof(n).new 6 
  return n | 1 == 3 if n < 5              # n: 0,1,4|false, 2,3|true 
  return false if n.gcd(6) != 1           # 1/3 (2/6) of integers are P3 pc
  p = typeof(n).new(5)                    # first P3 sequence value
  until p*p > n
    return false if n % p == 0 || n % (p + 2) == 0  # if n is composite
    p += 6      # first prime candidate for next kth residues group
  end
  true
end
```

When I always make n a ``BigInt`` the code compiles and runs correctly, with the line
``pn = typeof(n).new(6)`` commented out. However, when I uncomment that line and try to set n as the same Integer type as the input I get this error.

```
def primep3?(n)                           # P3 Prime Generator primality test
  # P3 = 6*k + {5, 7}                     # P3 primes candidates (pc) sequence
  #n = n.to_big_i 
  n = typeof(n).new 6 
  return n | 1 == 3 if n < 5              # n: 0,1,4|false, 2,3|true 
  return false if n.gcd(6) != 1           # 1/3 (2/6) of integers are P3 pc
  p = typeof(n).new(5)                    # first P3 sequence value
  until p*p > n
    return false if n % p == 0 || n % (p + 2) == 0  # if n is composite
    p += 6      # first prime candidate for next kth residues group
  end
  true
end


In primalitybytrialdivison.cr:13:17

 13 | n = typeof(n).new 6
                    ^--
Error: wrong number of arguments for '(Int32 | UInt64).new' (given 1, expected 0)
```
This makes no sense to me.
What am I doing wrong or is this a bug?

**Comment by z64** [comment]
To step through this, the full error is:

```
error in line 4
Error: wrong number of arguments for '(Int32 | UInt64).new' (given 1, expected 0)

Overloads are:
 - Union(T).new()
```

You can reproduce this without `typeof`, producing the same error:

```cr
struct Foo
end

struct Bar
end

def test(mixed : T) forall T
  T.new("something")
end

mixed = Foo.new || Bar.new
test(mixed)
```
```
error in line 8
Error: wrong number of arguments for '(Bar | Foo).new' (given 1, expected 0)

Overloads are:
 - Union(T).new()
```
This happens whether either type has a `T.new(string)` method or not.

If you do what the compiler suggests and call `Union(T).new()` with the above code instead:

```cr
def test(mixed : T) forall T
  T.new
end
```
You get this:

```
In /usr/lib/crystal/union.cr:19:1

 19 | struct Union
      ^-------
Error: can't create instance of a union type
```
Which seems like the actual problem you are running into; you can't create instances of unions directly. They can only be "created" by the compiler as the result of mixed-type expressions, and then instance methods can be dispatched across the union's types, i.e. `mixed.run` as long as `Foo` and `Bar` implement `run`.

You can fix your code by removing `typeof(n).new` and just writing `n = 6`, or as you stated, always converting `n` into a single type.

I don't know how accurate the error message is here - perhaps this is valid in some other context? - or whether it could be improved.

**Comment by straight-shoota** [comment]
TLDR; the key issue here is that `typeof(n)` returns *all* types `n` *can* have and is calculated at compile time. If that's more than one type, it returns a union. So `typeof(n).foo` calls `Union(*T)#foo`. In your example you try to call a method that doesn't exists. Everything works as expected.

You probably don't want to use `typeof(n)` here, but actually `n.class` which returns the *actual* type of `n` at runtime.
